### ISSUE TITLE
[BUG] Buggy behaviour on Raspberry pi 4 and Zwift

### ISSUE THREAD

**Issue by mazafra1** [cagnulein/qdomyos-zwift#2810]
**Describe the bug**
The Wahoo v5 is detected but nothing happens. 

**To Reproduce**
1. Launch program with `sudo ./qdomyos-zwift -no-gui -no-heart-service`
2. Try pairing on Zwift for iPad
3. Dircom device is not showing sometimes, other times shows, other times a residual `Wahoo KICKR 0000` appears. And only showing in Power source and Cadence, never in Resistance.
4. Bluetooth virtual device never shows up.

**Expected behavior**
It should show a Bluetooth Device to be paired in Zwift, to allow Zwift Click to be paired and then move Power source and Cadence to the virtual Dircon.

**Screenshots**
![Captura 2024-11-22 a las 9 41 01](https://github.com/user-attachments/assets/279d2ec8-e9b8-458e-ab81-10c86b429029)

**Desktop (please complete the following information):**
 - OS: `Linux raspberrypi 6.6.51+rpt-rpi-v8 #1 SMP PREEMPT Debian 1:6.6.51-1+rpt3 (2024-10-08) aarch64 GNU/Linux`
 - Version: source code from Nov 21st 2024.

**Smartphone (please complete the following information):**
N/A

**Append a debug log**
[debug-Fri_Nov_22_09_12_49_2024.log.zip](https://github.com/user-attachments/files/17867469/debug-Fri_Nov_22_09_12_49_2024.log.zip)
  
  
**Additional context**

**Comment by cagnulein** [comment]
the main issue is this one

```
Fri Nov 22 09:13:21 2024 1732263201122 Debug:   command 10 failed with status 12
Fri Nov 22 09:13:21 2024 1732263201122 Debug:   Advertising disable failed, ignoring
Fri Nov 22 09:13:21 2024 1732263201122 Debug:   sending command; ogf: 8 ocf: 6
Fri Nov 22 09:13:21 2024 1732263201122 Debug:   command sent successfully
Fri Nov 22 09:13:21 2024 1732263201124 Debug:   HCI event triggered, type: e
Fri Nov 22 09:13:21 2024 1732263201125 Debug:   command 6 executed successfully
Fri Nov 22 09:13:21 2024 1732263201125 Debug:   sending command; ogf: 8 ocf: 8
Fri Nov 22 09:13:21 2024 1732263201125 Debug:   command sent successfully
Fri Nov 22 09:13:21 2024 1732263201133 Debug:   HCI event triggered, type: 13
Fri Nov 22 09:13:21 2024 1732263201136 Debug:   HCI event triggered, type: e
Fri Nov 22 09:13:21 2024 1732263201136 Debug:   command 8 executed successfully
Fri Nov 22 09:13:21 2024 1732263201137 Debug:   sending command; ogf: 8 ocf: 9
Fri Nov 22 09:13:21 2024 1732263201137 Debug:   command sent successfully
Fri Nov 22 09:13:21 2024 1732263201143 Debug:   HCI event triggered, type: e
Fri Nov 22 09:13:21 2024 1732263201144 Debug:   command 9 executed successfully
Fri Nov 22 09:13:21 2024 1732263201144 Debug:   sending command; ogf: 8 ocf: 10
Fri Nov 22 09:13:21 2024 1732263201144 Debug:   command sent successfully
Fri Nov 22 09:13:21 2024 1732263201147 Debug:   HCI event triggered, type: e
Fri Nov 22 09:13:21 2024 1732263201148 Debug:   command 10 failed with status 12
Fri Nov 22 09:13:21 2024 1732263201148 Warning:   received advertising error
Fri Nov 22 09:13:21 2024 1732263201148 Debug: virtualdevices/virtualbike.cpp void virtualbike::error(QLowEnergyController::Error) "virtualbike::controller:ERROR AdvertisingError"
Fri Nov 22 09:13:21 2024 1732263201148 Debug: virtualdevices/virtualbike.cpp void virtualbike::reconnect() "virtualbike::reconnect"
```

try to investigate which could be the cause.
you can try to check if enabling the fake bike setting setting the manual device to wifi, at least show the virtual device on zwift
let me know

**Comment by mazafra1** [comment]
I launched it now with `sudo ./qdomyos-zwift -no-gui -no-heart-service -only-virtualbike` and I got: 
`Fri Nov 22 12:01:28 2024 1732273288841 Warning:   received advertising error
qt.bluetooth.bluez: received advertising error
Fri Nov 22 12:01:28 2024 1732273288841 Debug: virtualdevices/virtualbike.cpp void virtualbike::error(QLowEnergyController::Error) "virtualbike::controller:ERROR AdvertisingError"
"virtualbike::controller:ERROR AdvertisingError"
Fri Nov 22 12:01:28 2024 1732273288842 Debug: virtualdevices/virtualbike.cpp void virtualbike::reconnect() "virtualbike::reconnect"
"virtualbike::reconnect"
Fri Nov 22 12:01:28 2024 1732273288843 Warning:   Overriding existing local service with uuid "{00001826-0000-1000-8000-00805f9b34fb}"
Overriding existing local service with uuid "{00001826-0000-1000-8000-00805f9b34fb}"`

That's what you meant?

**Comment by cagnulein** [comment]
it seems there is a general error so. does it the same with the fake treadmill? if so it seems something on the raspberry itself
let me know

**Comment by mazafra1** [comment]
I reinstalled bluetooth kernel module and now it's working!!!! Weird. 

Now I have both the virtual Dircon and the virtual bluetooth QZPI and they're transmitting power. Is it possible to also get the zwift click in the same way the mobile app does?

**Comment by cagnulein** [comment]
weird!
add the -zwift_click switch to qz as well

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 22 nov 2024 alle ore 13:20 Mazafra ***@***.***>
ha scritto:

**Comment by mazafra1** [comment]
No effect whatsoever.

[debug-Fri_Nov_22_14_07_34_2024.log.zip](https://github.com/user-attachments/files/17870581/debug-Fri_Nov_22_14_07_34_2024.log.zip)

**Comment by cagnulein** [comment]
the flag is working but there are bluetooth issues

Fri Nov 22 14:07:51 2024 1732280871882 Debug:
zwift_play/zwiftclickremote.cpp void
zwiftclickremote::controllerStateChanged(QLowEnergyController::ControllerState)
"controllerStateChanged" QLowEnergyController::ConnectingState
Fri Nov 22 14:08:16 2024 1732280896787 Debug: devices/bluetooth.cpp void
bluetooth::debug(const QString&)
"zwiftclickremote::errorUnknownErrorUnknown Error"
Fri Nov 22 14:08:16 2024 1732280896789 Debug:
zwift_play/zwiftclickremote.cpp void
zwiftclickremote::controllerStateChanged(QLowEnergyController::ControllerState)
"controllerStateChanged" QLowEnergyController::UnconnectedState
Fri Nov 22 14:08:16 2024 1732280896789 Debug:
zwift_play/zwiftclickremote.cpp void
zwiftclickremote::controllerStateChanged(QLowEnergyController::ControllerState)
"trying to connect back again..."
Fri Nov 22 14:08:16 2024 1732280896840 Debug:
zwift_play/zwiftclickremote.cpp void
zwiftclickremote::controllerStateChanged(QLowEnergyController::ControllerState)
"controllerStateChanged" QLowEnergyController::ConnectingState
Fri Nov 22 14:08:16 2024 1732280896848 Debug: devices/bluetooth.cpp void
bluetooth::debug(const QString&)
"zwiftclickremote::errorUnknownErrorUnknown Error"

maybe there are pending issues on bluetooth on your raspberry?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 22 nov 2024 alle ore 14:12 Mazafra ***@***.***>
ha scritto:

**Comment by mazafra1** [comment]
Thanks Roberto. I've reviewed and I see nothing off except the error above. 

Is there any test I could do to troubleshoot the Zwift Click only?

**Comment by mazafra1** [comment]
Yes, it can connect and successfully retrieved these characteristics:
```
Primary Service (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0026
        0000180f-0000-1000-8000-00805f9b34fb
        Battery Service
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0026/char0027
        00002a19-0000-1000-8000-00805f9b34fb
        Battery Level
Descriptor (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0026/char0027/desc0029
        00002902-0000-1000-8000-00805f9b34fb
        Client Characteristic Configuration
Primary Service (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0017
        00000001-19ca-4651-86e5-fa29dcdd09d1
        Vendor specific
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0017/char0023
        00000006-19ca-4651-86e5-fa29dcdd09d1
        Vendor specific
Descriptor (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0017/char0023/desc0025
        00002902-0000-1000-8000-00805f9b34fb
        Client Characteristic Configuration
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0017/char0020
        00000005-19ca-4651-86e5-fa29dcdd09d1
        Vendor specific
Descriptor (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0017/char0020/desc0022
        00002902-0000-1000-8000-00805f9b34fb
        Client Characteristic Configuration
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0017/char001d
        00000004-19ca-4651-86e5-fa29dcdd09d1
        Vendor specific
Descriptor (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0017/char001d/desc001f
        00002902-0000-1000-8000-00805f9b34fb
        Client Characteristic Configuration
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0017/char001b
        00000003-19ca-4651-86e5-fa29dcdd09d1
        Vendor specific
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0017/char0018
        00000002-19ca-4651-86e5-fa29dcdd09d1
        Vendor specific
Descriptor (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service0017/char0018/desc001a
        00002902-0000-1000-8000-00805f9b34fb
        Client Characteristic Configuration
Primary Service (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service000e
        0000180a-0000-1000-8000-00805f9b34fb
        Device Information
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service000e/char0015
        00002a26-0000-1000-8000-00805f9b34fb
        Firmware Revision String
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service000e/char0013
        00002a27-0000-1000-8000-00805f9b34fb
        Hardware Revision String
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service000e/char0011
        00002a25-0000-1000-8000-00805f9b34fb
        Serial Number String
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service000e/char000f
        00002a29-0000-1000-8000-00805f9b34fb
        Manufacturer Name String
Primary Service (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service000a
        00001801-0000-1000-8000-00805f9b34fb
        Generic Attribute Profile
Characteristic (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service000a/char000b
        00002a05-0000-1000-8000-00805f9b34fb
        Service Changed
Descriptor (Handle 0x0000)
        /org/bluez/hci0/dev_DC_30_32_2E_9B_0F/service000a/char000b/desc000d
        00002902-0000-1000-8000-00805f9b34fb
        Client Characteristic Configuration
```

Should I try subscribing to any one in particular?

**Comment by cagnulein** [comment]
so it could be the issue is when you are connecting to both trainer and the click? can you try to connect to both?
first connect to the trainer, and then try to subscribe to the 0x2AD2 for example and then connect to the click. let's see if it works

**Comment by mazafra1** [comment]
I tried that and it all seems OK. I even subscribed to notifications on `0x2AD2`. Although I'm not sure if anything should happen. But every step threw a success message.

**Comment by cagnulein** [comment]
OK it could be something about the fact that qz is trying to also
broadcasting over Bluetooth at the same time. Disable the virtual Bluetooth
device from the qz settings so we should be on the same Bluetoothctl
context

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 1 dic 2024 alle 09:35 Mazafra ***@***.***> ha
scritto:

**Comment by mazafra1** [comment]
Thanks. That makes sense. What’s the terminal  parameter for that?

Cheers

**Comment by cagnulein** [comment]
i created this

"-no-virtual-device-bluetooth"

build qz from the master, i just added it for you

**Comment by mazafra1** [comment]
Thanks a ton!! Now it works!!! Obviously without virtual bluetooth device.
![IMG_0248](https://github.com/user-attachments/assets/6cacd7b7-1eb0-4011-aa35-62f0de599b66)

What can we do then now? Is the problem the number of bt channels occupied or something else? Did you experience something similar on your Rpi zero?

**Comment by cagnulein** [comment]
there are 2 ways, check here https://github.com/cagnulein/qdomyos-zwift/wiki/Virtual-Gearing-on-QZ-with-Zwift,-Rouvy,-Mywhoosh-and-Traning-Peaks

if you use the method with qz directly, you can avoid to use the bluetooth interface on zwift so it would already work with the actually raspberry version (the only downside is that you will not see the gears on zwift)

**Comment by cagnulein** [comment]
yes it would work the same, since now qz has MQTT you can display the gears on home assistant for example!

**Comment by mazafra1** [comment]
Thanks. How do I need to configure the screen below?
![image](https://github.com/user-attachments/assets/e57a0e77-e2e3-42ff-8340-3d1eba078009)

**Comment by cagnulein** [comment]
you can leave as it is, as i said, without the bluetooth interface we can't have the gears from/to zwift since it's bluetooth based

**Comment by cagnulein** [comment]
I guess it's what I fixed here @mazafra1 

https://github.com/cagnulein/qdomyos-zwift/issues/2892
